Google Maps shortcode AMP compatibility fix: external links

kienstra opened the pull request:
#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* In the Google Maps shortcode, allow clicking external links on AMP endpoints

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This adds functionality to an existing feature 

#### Testing instructions:
1. Ensure the [AMP plugin](https://wordpress.org/plugins/amp/) is enabled
2. In AMP Settings, select 'Transitional' mode
3. In Jetpack > Settings, ensure 'Compose using shortcodes' is toggled on:


![compose-shortcodes](https://user-images.githubusercontent.com/4063887/78590154-fee85f80-7806-11ea-8291-b15b586a7ae3.png)

4. Add a Shortcode block with:
```html
[googlemaps https://maps.google.com/maps?f=q&amp;hl=en&amp;geocode=&amp;q=San+Francisco,+CA&amp;sll=43.469466,-83.998504&amp;sspn=0.01115,0.025942&amp;g=San+Francisco,+CA&amp;ie=UTF8&amp;z=12&amp;iwloc=addr&amp;ll=37.808156,-122.402458&amp;output=embed&amp;s=AARTsJp56EajYksz3JXgNCwT3LJnGsqqAQ&amp;w=425&amp;h=350]
```
(this was taken from [googlemaps.php](https://github.com/Automattic/jetpack/blob/30f4c13f59bfeae1eeb2d03e1ad3e81c90feeb54/modules/shortcodes/googlemaps.php#L7))
5. Preview the AMP front-end by clicking the AMP icon:

![amp-button](https://user-images.githubusercontent.com/4063887/78590589-aebdcd00-7807-11ea-843e-625fd0a5c00c.png)

6. Click an external link, like 'View larger map'
7. Expected: this opens a new tab or window:
![external-link](https://user-images.githubusercontent.com/4063887/78591253-c6e21c00-7808-11ea-9c61-935a8143aa1e.gif)


Before, clicking an external link did nothing.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Improve Google Maps shortcode AMP compatibility

kienstra committed:
In AMP, allow opening external links Google Maps

Add sandbox="allow-popups",
in addition to other sandbox values,
which are defaults:
https://github.com/ampproject/amp-wp/blob/d01879156199df5dcf60809a353a077123d75e54/includes/sanitizers/class-amp-iframe-sanitizer.php#L25

kienstra commented on modules/shortcodes/googlemaps.php:
```diff
@@ -124,7 +124,9 @@ function jetpack_googlemaps_shortcode( $atts ) {
 		}
 
-		return '<div class="' . esc_attr( $css_class ) . '"><iframe width="' . $width . '" height="' . $height . '" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" src="' . $url . '"></iframe></div>';
+		$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ? 'sandbox="allow-popups allow-scripts allow-same-origin"' : '';
```
In the AMP plugin, `sandbox="allow-scripts allow-same-origin"` [is the default](https://github.com/ampproject/amp-wp/blob/c12acec0260a0f22401d115cfdd84af63b717816/includes/sanitizers/class-amp-iframe-sanitizer.php#L23) for `sandbox`. So this also includes those.

`sandbox="allow-popups"` is the fix for this PR. It allows opening external links.

kienstra commented:
Hi @MattGeri,
Could you please review this? 

Thanks, Matt!

jeherve commented on modules/shortcodes/googlemaps.php:
```diff
@@ -124,7 +124,9 @@ function jetpack_googlemaps_shortcode( $atts ) {
-		return '<div class="' . esc_attr( $css_class ) . '"><iframe width="' . $width . '" height="' . $height . '" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" src="' . $url . '"></iframe></div>';
+		$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ? 'sandbox="allow-popups allow-scripts allow-same-origin"' : '';
+
+		return '<div class="' . esc_attr( $css_class ) . '"><iframe width="' . $width . '" height="' . $height . '" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" src="' . $url . '" ' . $sandbox . '></iframe></div>';
```
Should we take that opportunity to clean this up, escape everything, and make it more readable?

```diff
diff --git a/modules/shortcodes/googlemaps.php b/modules/shortcodes/googlemaps.php
index 6171b2241..c64ff8196 100644
--- a/modules/shortcodes/googlemaps.php
+++ b/modules/shortcodes/googlemaps.php
@@ -124,9 +124,20 @@ function jetpack_googlemaps_shortcode( $atts ) {
 			}
 		}
 
-		$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ? 'sandbox="allow-popups allow-scripts allow-same-origin"' : '';
-
-		return '<div class="' . esc_attr( $css_class ) . '"><iframe width="' . $width . '" height="' . $height . '" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" src="' . $url . '" ' . $sandbox . '></iframe></div>';
+		$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request()
+			? 'sandbox="allow-popups allow-scripts allow-same-origin"'
+			: '';
+
+		return sprintf(
+			'<div class="%1$s">
+				<iframe width="%2$d" height="%3$d" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" %5$s src="%4$s"></iframe>
+			</div>',
+			esc_attr( $css_class ),
+			absint( $width ),
+			absint( $height ),
+			esc_url( $url ),
+			esc_attr( $sandbox )
+		);
 	}
 }
 add_shortcode( 'googlemaps', 'jetpack_googlemaps_shortcode' );

```

kienstra replied:
Yeah, that's a good idea. I'll commit that later tonight if that's alright.

kienstra replied:
It's applied in https://github.com/Automattic/jetpack/pull/15317/commits/7aae669ff32ceb23268adcffba7b08a902e291e9

jeherve reviewed (changes requested):
This works well for me. I only have one nitpick, while we're editing that file.

kienstra committed:
Commit Jeremy's suggestion to clean up shortcode callback

Thanks to Alain Schlesser for the logic
that enables testing the HTML:
https://github.com/ampproject/amp-wp/blame/1c5d7b23b018142d58caea37c223facf17953905/lib/optimizer/tests/src/MarkupComparison.php#L21-L24

jeherve reviewed (approved):
This seems to test well for me. 👍

jeherve commented:
r206418-wpcom

westonruter commented on modules/shortcodes/googlemaps.php:
```diff
@@ -124,7 +124,20 @@ function jetpack_googlemaps_shortcode( $atts ) {
 		}
 
-		return '<div class="' . esc_attr( $css_class ) . '"><iframe width="' . $width . '" height="' . $height . '" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" src="' . $url . '"></iframe></div>';
+		$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request()
+			? 'sandbox="allow-popups allow-scripts allow-same-origin"'
+			: '';
```
Aside: this `sandbox` attribute would probably make sense in non-AMP as well.
